File: modules/spread/app/model/SpreadCorrector.java
package model;

import java.nio.file.Path;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public abstract class SpreadCorrector<D, S, C, F, ColorType> {

  public abstract void closeDocument(D compareDocument);

  public abstract String compareCellFormulas(C masterCell, C compareCell);

  public abstract String compareCellValues(C masterCell, C compareCell);

  public abstract String compareChartsInSheet(S compareSheet, S sampleSheet);

  public abstract String compareNumberOfChartsInDocument(D compareDocument, D sampleDocument);

  public abstract void compareSheet(S sampleTable, S compareTable, boolean conditionalFormating);

  public SpreadSheetCorrectionResult correct(Path samplePath, Path comparePath, boolean conditionalFormating,
      boolean compareCharts) throws CorrectionException {

    // Check if both documents exist as files
    if (!samplePath.toFile().exists())
      return new SpreadSheetCorrectionResult(false, Arrays.asList(StringConsts.ERROR_MISSING_SAMPLE));

    if (!comparePath.toFile().exists())
      return new SpreadSheetCorrectionResult(false, Arrays.asList(StringConsts.ERROR_MISSING_SOLUTION));

    // Load document, if loading returns null, return Error
    D sampleDocument = loadDocument(samplePath);
    if (sampleDocument == null)
      return new SpreadSheetCorrectionResult(false, Arrays.asList(StringConsts.ERROR_LOAD_SAMPLE));

    D compareDocument = loadDocument(comparePath);
    if (compareDocument == null)
      return new SpreadSheetCorrectionResult(false, Arrays.asList(StringConsts.ERROR_LOAD_SOLUTION));

    if (getSheetCount(sampleDocument) != getSheetCount(compareDocument))
      return new SpreadSheetCorrectionResult(false, Arrays.asList(StringConsts.ERROR_WRONG_SHEET_NUM));

    if (compareCharts) {
      String message = compareNumberOfChartsInDocument(compareDocument, sampleDocument);
      // write message in Cell A0 on first Sheet
      setCellComment(getCellByPosition(getSheetByIndex(compareDocument, 0), 0, 0), message);
    }

    LinkedList<String> notices = new LinkedList<>();

    // Iterate over sheets
    int sheetCount = getSheetCount(sampleDocument);
    for (int sheetIndex = 0; sheetIndex < sheetCount; sheetIndex++) {
      S sampleTable = getSheetByIndex(sampleDocument, sheetIndex);
      S compareTable = getSheetByIndex(compareDocument, sheetIndex);
      if (compareTable == null || sampleTable == null)
	notices.add(String.format(StringConsts.ERROR_LOAD_TABLE_VAR, (sheetCount + 1)));
      else
	compareSheet(sampleTable, compareTable, conditionalFormating);
    }

    // Save and close workbooks
    saveCorrectedSpreadsheet(compareDocument, comparePath);
    closeDocument(compareDocument);
    closeDocument(sampleDocument);

    if (notices.isEmpty())
      return new SpreadSheetCorrectionResult(true, Arrays.asList(StringConsts.SUCCESS_CORRECTION));
    else
      return new SpreadSheetCorrectionResult(false, notices);
  }

  public abstract C getCellByPosition(S table, int row, int column);

  public abstract List<C> getColoredRange(S master);

  public abstract S getSheetByIndex(D sampleDocument, int sheetIndex);

  public abstract int getSheetCount(D sampleDocument);

  /**
   * Loads a document from a given path
   *
   * @param musterPath
   *          - path to the document
   * @return the document if there is a document that can be loaded, else null
   */
  public abstract D loadDocument(Path musterPath) throws CorrectionException;

  public abstract void saveCorrectedSpreadsheet(D compareDocument, Path testPath);

  public abstract void setCellComment(C cell, String comment);

  public abstract void setCellStyle(C cell, F font, ColorType color);

}

File: app/model/spread/SpreadCorrector.scala
package model.spread

import java.nio.file.Path

import model.core.FileUtils
import model.core.result.{EvaluationResult, SuccessType}
import model.spread.SpreadConsts._

import scala.collection.mutable.ListBuffer
import scala.util.{Failure, Success, Try}

// Types of results

abstract sealed class SpreadSheetCorrectionResult(val success: SuccessType, val notices: Seq[String]) extends EvaluationResult

final case class SpreadSheetCorrectionFailure(cause: String) extends SpreadSheetCorrectionResult(SuccessType.ERROR, Seq(cause))

final case class SpreadSheetCorrectionError(strs: Seq[String]) extends SpreadSheetCorrectionResult(SuccessType.NONE, strs)

case object SpreadSheetCorrectionSuccess extends SpreadSheetCorrectionResult(SuccessType.COMPLETE, Seq(SUCCESS_CORRECTION))

// Actual corrector

abstract class SpreadCorrector extends FileUtils {

  // Abstract types

  protected type DocumentType

  protected type SheetType

  protected type CellType

  protected type FontType

  protected type ColorType

  // Colors

  protected val green: ColorType

  protected val red: ColorType


  def correct(samplePath: Path, comparePath: Path, conditionalFormating: Boolean, compareCharts: Boolean): SpreadSheetCorrectionResult = {
    if (!samplePath.toFile.exists) SpreadSheetCorrectionFailure(ERROR_MISSING_SAMPLE)

    else if (!comparePath.toFile.exists) SpreadSheetCorrectionFailure(ERROR_MISSING_SOLUTION)

    else loadDocument(samplePath) match {
      case Failure(_) => SpreadSheetCorrectionFailure(ERROR_LOAD_SAMPLE)

      case Success(sampleDocument) => loadDocument(comparePath) match {

        case Failure(_) => SpreadSheetCorrectionFailure(ERROR_LOAD_SOLUTION)

        case Success(compareDocument) =>
          if (getSheetCount(sampleDocument) != getSheetCount(compareDocument))
            SpreadSheetCorrectionFailure(ERROR_WRONG_SHEET_NUM)
          else {
            if (compareCharts) {
              val (_, chartCompMessage) = compareNumberOfChartsInDocument(compareDocument, sampleDocument)
              // write message in RobotCell A0 on first Sheet
              getCellByPosition(getSheetByIndex(compareDocument, 0), 0, 0) match {
                case None       => Unit
                case Some(cell) => setCellComment(cell, chartCompMessage)
              }
            }

            val notices: ListBuffer[String] = ListBuffer.empty

            // Iterate over sheets
            val sheetCount = getSheetCount(sampleDocument)
            for (sheetIndex <- 0 until sheetCount) {

              val sampleTable: SheetType = getSheetByIndex(sampleDocument, sheetIndex)
              val compareTable: SheetType = getSheetByIndex(compareDocument, sheetIndex)

              if (compareTable == null || sampleTable == null)
                notices += s"""Beim Öffnen der ${sheetCount + 1}. Tabelle ist ein Fehler aufgetreten."""
              else
                compareSheet(sampleTable, compareTable, conditionalFormating)
            }

            // Save and close workbooks
            saveCorrectedSpreadsheet(compareDocument, comparePath)
            closeDocument(compareDocument)
            closeDocument(sampleDocument)

            if (notices.isEmpty) SpreadSheetCorrectionSuccess
            else SpreadSheetCorrectionError(notices.toList)
          }
      }
    }
  }


  protected def compareCellFormulas(masterCell: CellType, compareCell: CellType): (Boolean, String)

  protected def compareCellValues(masterCell: CellType, compareCell: CellType): (Boolean, String)

  protected def compareChartsInSheet(compareSheet: SheetType, sampleSheet: SheetType): (Boolean, String)

  protected def compareNumberOfChartsInDocument(compareDocument: DocumentType, sampleDocument: DocumentType): (Boolean, String)

  protected def compareSheet(sampleTable: SheetType, compareTable: SheetType, conditionalFormating: Boolean): Unit

  protected def compareSheetConditionalFormatting(master: SheetType, compare: SheetType): Seq[String]

  // Getters and other helper methods

  protected def getCellByPosition(table: SheetType, row: Int, column: Int): Option[CellType]

  protected def getColoredRange(master: SheetType): Seq[CellType]

  protected def getSheetByIndex(sampleDocument: DocumentType, sheetIndex: Int): SheetType

  protected def getSheetCount(sampleDocument: DocumentType): Int

  protected def closeDocument(compareDocument: DocumentType): Unit

  protected def loadDocument(musterPath: Path): Try[DocumentType]

  protected def saveCorrectedSpreadsheet(compareDocument: DocumentType, testPath: Path): Try[Path]

  protected def setCellComment(cell: CellType, comment: String): Unit

  protected def setCellStyle(cell: CellType, font: FontType, success: Boolean): Unit

}
